compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -244,22 +244,17 @@ pub fn each_linked_rlib(
 
         fmts
     } else {
-        for combination in info.dependency_formats.iter().combinations(2) {
-            let (ty1, list1) = &combination[0];
-            let (ty2, list2) = &combination[1];
-            if list1 != list2 {
-                return Err(errors::LinkRlibError::IncompatibleDependencyFormats {
-                    ty1: format!("{ty1:?}"),
-                    ty2: format!("{ty2:?}"),
-                    list1: format!("{list1:?}"),
-                    list2: format!("{list2:?}"),
-                });
-            }
-        }
-        if info.dependency_formats.is_empty() {
-            return Err(errors::LinkRlibError::MissingFormat);
+        let mut dep_formats = info.dependency_formats.iter();
+        let (ty1, list1) = dep_formats.next().ok_or(errors::LinkRlibError::MissingFormat)?;
+        if let Some((ty2, list2)) = dep_formats.find(|(_, list2)| list1 != *list2) {
```

Comment:
I don't think this is equivalent, because if say the crate type 2 and 3 disagree it will no longer emit an error.

Reply:
Is the `==` operator not transitive for this type?

Reply:
It is. But say the dependency formats have 3 elements (where each element is a list), where the the second list and the last list are different. Previously, this was detected, because all combinations are checked, but now it is no longer detected because you only compare the first list against all other lists. The second and third lists are no longer compared.

Reply:
I do agree with @kadiwa4 on this. In your example if the second and last lists are different, then the first and last list will also be different, and it would still be caught.
But I might just be completely wrong on this, I dunno 😅

Reply:
(assuming the second and first lists don't compare as equal)

Reply:
Yeah, that's what I thought.
This is the (slightly shortened) definition of the transitivity property from the std docs:
> `a == b` and `b == c` implies `a == c`

Now say `a = dep_formats[1].1`, `b = dep_formats[0].1` and `c = dep_formats[2].1`. In the `.find()` method, the code checks that `b == a` and `b == c`. Assuming that `find` finishes without finding a mismatch, we now know that according to the transitivity property, `a == c` is true (i.e. `dep_formats[1].1 == dep_formats[2].1`).
(I am assuming that `b == a` is the same as `a == b` of course.)

Reply:
Oh you're right, I was confused in the weeds here. If all elements are equal to the first one, then of course all the elements have to be equal!

Reply:
😁